llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1250,6 +1265,158 @@ void AddressSanitizerPass::printPipeline(
+            continue;
+
+          if (F.getName().equals("_Z12get_local_idj") ||
+              F.getName().equals("_Z14get_local_sizej") ||
+              F.getName().equals("_Z14_get_num_groupsj"))
+            continue;
+        }
+
+      } else {
+        if (F.isDeclaration() || F.getName().starts_with("__sycl_") ||
+            F.getName().starts_with("__spirv") ||
+            F.getFunctionType()->isVarArg())
+          continue;
+      }
+      SpirFixupFuncs.emplace_back(&F);
+    }
+  }
+
+  SmallVector<std::pair<Function *, Function *>> SpirFuncs;
+  const int LongSize = M.getDataLayout().getPointerSizeInBits();
+  auto *IntptrTy = Type::getIntNTy(M.getContext(), LongSize);
+
+  for (auto *F : SpirFixupFuncs) {
+    SmallVector<Type *, 16> Types;
+    for (Function::const_arg_iterator I = F->arg_begin(), E = F->arg_end();
+         I != E; ++I) {
+      Types.push_back(I->getType());
+    }
+
+    // New Argument
+    Types.push_back(IntptrTy);
+
+    FunctionType *NewFTy = FunctionType::get(F->getReturnType(), Types, false);
+
+    std::string OrigFuncName = F->getName().str();
+    F->setName(OrigFuncName + "_del");
+
+    Function *NewF =
+        Function::Create(NewFTy, F->getLinkage(), OrigFuncName, F->getParent());
+    NewF->copyAttributesFrom(F);
+    NewF->copyMetadata(F, 0);
+    NewF->setCallingConv(F->getCallingConv());
+    NewF->setDSOLocal(F->isDSOLocal());
+
+    // Set original arguments' names.
+    Function::arg_iterator NewI = NewF->arg_begin();
+    for (Function::const_arg_iterator I = F->arg_begin(), E = F->arg_end();
+         I != E; ++I, ++NewI) {
+      NewI->setName(I->getName());
+    }
+    // New argument name
+    NewI->setName("__asan_launch");
+
+    NewF->splice(NewF->begin(), F);
+    assert(F->isDeclaration() &&
+           "splice does not work, original function body is not empty!");
+
+    NewF->setSubprogram(F->getSubprogram());
+
+    NewF->setComdat(F->getComdat());
+    F->setComdat(nullptr);
+
+    F->deleteBody();
+
+    for (Function::arg_iterator I = F->arg_begin(), E = F->arg_end(),
+                                NI = NewF->arg_begin();
+         I != E; ++I, ++NI) {
+      I->replaceAllUsesWith(&*NI);
+    }
+
+    // Fixup metadata
+    IRBuilder<> Builder(M.getContext());
+
+    auto FixupMetadata = [&NewF](StringRef MDName, Constant *NewV) {
+      auto *Node = NewF->getMetadata(MDName);
+      if (!Node)
+        return;
+      SmallVector<Metadata *, 8> NewMD;
+      for (unsigned I = 0; I < Node->getNumOperands(); ++I) {
+        NewMD.emplace_back(Node->getOperand(I));
+      }
+      NewMD.emplace_back(ConstantAsMetadata::get(NewV));
+      NewF->setMetadata(MDName, llvm::MDNode::get(NewF->getContext(), NewMD));
+    };
+
+    FixupMetadata("kernel_arg_buffer_location", Builder.getInt32(-1));
+    FixupMetadata("kernel_arg_runtime_aligned", Builder.getFalse());
+    FixupMetadata("kernel_arg_exclusive_ptr", Builder.getFalse());
```

Comment:
are metadata `!kernel_arg_addr_space  !kernel_arg_access_qual  !kernel_arg_type  !kernel_arg_base_type  !kernel_arg_type_qual  !kernel_arg_accessor_ptr` fixed?

Reply:
Offline sync with @wenju-he, these metadata don't needed to update here.

Reply:
These metadata probably still need updating as I've met a case that fails because these metadata are not updated by AddressSanitizerPass